src/platform/test/mtest-main.c

```diff
@@ -0,0 +1,215 @@
+static bool _parseMTestOpts(struct mSubParser* parser, int option, const char* arg);
+static enum mRegisterExitCode _parseNamedRegister(const char* regStr);
+
+static bool _dispatchExiting = false;
+
+#ifdef M_CORE_GBA
+void (*_armSwi16)(struct ARMCore* cpu, int immediate);
+void (*_armSwi32)(struct ARMCore* cpu, int immediate);
+
+static void _mTestSwi16(struct ARMCore* cpu, int immediate);
+static void _mTestSwi32(struct ARMCore* cpu, int immediate);
+static void _mTestUpdateExitCond(struct ARMCore* cpu, int immediate);
+
+static enum mRegisterExitCode _returnCodeRegister;
+static int _exitCode = 0;
+static int _prevSwiImmediate = -1;
+#endif
+
+int main(int argc, char * argv[]) {
+	signal(SIGINT, _mTestShutdown);
+
+	struct MTestOpts mTestOpts = { -1, mRegister_NONE };
+	struct mSubParser subparser = {
+		.usage = MTEST_USAGE,
+		.parse = _parseMTestOpts,
+		.extraOptions = MTEST_OPTIONS,
+		.opts = &mTestOpts
+	};
+
+	struct mArguments args;
+	bool parsed = parseArguments(&args, argc, argv, &subparser);
+	if (!args.fname) {
+		parsed = false;
+	}
+	if (!parsed || args.showHelp) {
+		usage(argv[0], MTEST_USAGE);
+		return !parsed;
+	}
+	if (args.showVersion) {
+		version(argv[0]);
+		return 0;
+	}
+	struct mCore* core = mCoreFind(args.fname);
+	if (!core) {
+		return 1;
+	}
+	core->init(core);
+	mCoreInitConfig(core, "mTest");
+	applyArguments(&args, NULL, &core->config);
+
+	mCoreConfigSetDefaultValue(&core->config, "idleOptimization", "remove");
+
+#ifdef M_CORE_GBA
+	if (core->platform(core) == mPLATFORM_GBA) {
+		((struct GBA*) core->board)->hardCrash = false;
+
+		_armSwi16 = ((struct GBA*) core->board)->cpu->irqh.swi16;
+		((struct GBA*) core->board)->cpu->irqh.swi16 = _mTestSwi16;
+		_armSwi32 = ((struct GBA*) core->board)->cpu->irqh.swi32;
+		((struct GBA*) core->board)->cpu->irqh.swi32 = _mTestSwi32;
+
+		_returnCodeRegister = mTestOpts.returnCodeRegister;
+	}
+#endif
+
+	bool cleanExit = true;
+	if (!mCoreLoadFile(core, args.fname)) {
+		cleanExit = false;
+		goto loadError;
+	}
+	if (args.patch) {
+		core->loadPatch(core, VFileOpen(args.patch, O_RDONLY));
+	}
+
+	struct VFile* savestate = NULL;
+
+	if (args.savestate) {
+		savestate = VFileOpen(args.savestate, O_RDONLY);
+	}
+
+	core->reset(core);
+
+	struct mCheatDevice* device;
+	if (args.cheatsFile && (device = core->cheatDevice(core))) {
+		struct VFile* vf = VFileOpen(args.cheatsFile, O_RDONLY);
+		if (vf) {
+			mCheatDeviceClear(device);
+			mCheatParseFile(device, vf);
+			vf->close(vf);
+		}
+	}
+
+	if (savestate) {
+		mCoreLoadStateNamed(core, savestate, 0);
+		savestate->close(savestate);
+		savestate = 0;
+	}
+
+	int returnCode = _mTestRunloop(core, mTestOpts.exitSwiImmediate);
+
+	core->unloadROM(core);
+
+loadError:
+	freeArguments(&args);
+	mCoreConfigDeinit(&core->config);
+	core->deinit(core);
+
+	return cleanExit ? returnCode : 1;
+}
+
+static int _mTestRunloop(struct mCore* core, int exitSwiImmediate) {
+	do {
+		core->step(core);
```

Comment:
```suggestion
		core->runLoop(core);
```
Will be much faster, and an SWI call should always break out of the runloop, especially one to GBAStop.